### PR TITLE
[FSDP2] add bfloat16 into training.mixed_precision_reduce

### PR DESCRIPTION
**Pull request by mori360**
Fix https://github.com/pytorch/torchtitan/issues/1027
Add choice of "bfloat16" to allow users more choices in mixed_precision_reduce.
"bfloat16" in `mixed_precision_reduce might` might affect the performance as it's mixed_precision at gradient, we keep "float32" as default but add one more choice

mixed_precision_param = "bfloat16", mixed_precision_reduce = "float32"
Screenshot 2025-03-28 at 2 27 00 PM

mixed_precision_param = "bfloat16", mixed_precision_reduce = "bfloat16"
Screenshot 2025-03-28 at 2 27 07 PM

### CODE REVIEW

**Review by lessw2020 — APPROVED** [review]
cool, lgtm.
generic question but any plans to run with both dtypes under deterministic and compare the loss curves to advertise this change?

**Comment by mori360** [comment]
> cool, lgtm. generic question but any plans to run with both dtypes under deterministic and compare the loss curves to advertise this change?

Thanks for the comment, let me add them before landing.

**Comment by lkhphuc** [comment]
I saw slightly higher tflops and tps, likely indicating less time for communication. But the memory usage is also slightly higher, which is counterintuitive since we are using less precision. 
Do you have any idea why?

**Comment by mori360** [comment]
> I saw slightly higher tflops and tps, likely indicating less time for communication. But the memory usage is also slightly higher, which is counterintuitive since we are using less precision. Do you have any idea why?

That's expected.
For mix_precision_reduce with bfloat16, it would cast params to bfloat16 during forward and backward, causing higher memory cost.
Since we cast params from float32 to bloat16, there's less communication cost due to low precision.
